Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,207 @@
+/-
+Copyright (c) 2023 Adrian Wüthrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrian Wüthrich
+-/
+import Mathlib.Combinatorics.SimpleGraph.AdjMatrix
+import Mathlib.Combinatorics.SimpleGraph.IncMatrix
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.IsDiag
+
+/-!
+# Laplacian Matrix
+
+This module defines the Laplacian matrix of a graph, and proves some of its elementary properties.
+
+## Main definitions & Results
+
+* `SimpleGraph.degMatrix`: The degree matrix of a simple graph
+* `SimpleGraph.lapMatrix `: The Laplacian matrix of a simple graph, defined as the difference
+  between the degree matrix and the adjacency matrix.
+* `isPosSemidef_lapMatrix`: The Laplacian matrix is positive semidefinite.
+* `rank_ker_lapMatrix_eq_card_ConnectedComponent`: The number of connected components in `G` is
+  the dimension of the nullspace of its Laplacian matrix.
+
+-/
+
+
+open BigOperators Finset Matrix SimpleGraph
+
+variable {V α : Type*} (α) [Fintype V] [DecidableEq V] (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- The diagonal matrix consisting of the degrees of the vertices in the graph. -/
+def SimpleGraph.degMatrix [AddMonoidWithOne α] : Matrix V V α := Matrix.diagonal (G.degree ·)
+
+/-- `lapMatrix G R` is the matrix `L = D - A` where `D`is the degree
+  and `A` the adjacency matrix of `G`. -/
+def SimpleGraph.lapMatrix [AddGroupWithOne α] : Matrix V V α := G.degMatrix α - G.adjMatrix α
+
+variable {α}
+
+theorem isSymm_lapMatrix [AddGroupWithOne α] : (G.lapMatrix α).IsSymm := by
+  rw [Matrix.IsSymm, lapMatrix, degMatrix, transpose_sub, diagonal_transpose, transpose_adjMatrix]
+
+theorem lapMatrix_mulVec_const_eq_zero [Ring α] : mulVec (G.lapMatrix α) (fun _ ↦ 1) = 0 := by
+  ext1 i
+  simp_rw [lapMatrix, sub_mulVec, Pi.sub_apply, adjMatrix_mulVec_apply, degMatrix, mulVec,
+    dotProduct, diagonal]
+  simp
+
+theorem dotProduct_mulVec_adjMatrix [Ring α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.adjMatrix α) x = ∑ i : V, ∑ j : V, if G.Adj i j then x i * x j else 0 := by
+  unfold mulVec dotProduct
+  simp [mul_sum]
+
+theorem dotProduct_mulVec_degMatrix [CommRing α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.degMatrix α) x = ∑ i : V, G.degree i * x i * x i := by
+  unfold dotProduct degMatrix
+  simp only [mulVec_diagonal, ← mul_assoc, mul_comm (x _)]
+
+variable (α)
+
+theorem sum_adj_eq_degree [AddCommMonoidWithOne α] (i : V) :
```

Comment:
Given the theorem statement, it would be in the reverse order. Maybe throw in an `if`? So `degree_eq_sum_if_adj`